amethyst_assets/src/storage.rs

```diff
@@ -112,46 +118,133 @@ impl<A: Asset> AssetStorage<A> {
+                        name,
+                    } => {
+                        let (asset, reload_obj) = data.map(
+                            |FormatValue { data, reload }| (data, reload),
+                        ).and_then(|(d, rel)| f(d).map(|a| (a, rel)))
+                            .map_err(|e| AssetError::new(name, format, e))?;
+
+                        let id = handle.id();
+                        bitset.add(id);
+                        handles.push(handle.clone());
+
+                        // NOTE: the loader has to ensure that a handle will be used
+                        // together with a `Data` only once.
+                        unsafe {
+                            assets.insert(id, asset);
+                        }
+
+                        (reload_obj, handle)
+                    }
+                    Processed::HotReload {
+                        data,
+                        format,
+                        handle,
+                        name,
+                        old_reload,
+                    } => {
+                        let (asset, reload_obj) = match data.map(
+                            |FormatValue { data, reload }| (data, reload),
+                        ).and_then(|(d, rel)| f(d).map(|a| (a, rel)))
+                            .map_err(|e| AssetError::new(name, format, e))
+                        {
+                            Ok(x) => x,
+                            Err(e) => {
+                                eprintln!("Failed to hot-reload: {}", e);
+
+                                reloads.push((handle.downgrade(), old_reload));
+
+                                return Ok(());
+                            }
+                        };
+
+                        let id = handle.id();
+                        assert!(bitset.contains(id));
+                        unsafe {
+                            let old = assets.get_mut(id);
+                            *old = asset;
+                        }
+
+                        (reload_obj, handle)
+                    }
+                };
+
+                // Add the reload obj if it is `Some`.
+                if let Some(reload_obj) = reload_obj {
+                    reloads.push((handle.downgrade(), reload_obj));
                 }
 
                 Ok(())
             });
         }
 
-        while let Some(i) = self.handles.iter().position(Handle::is_unused) {
-            let old = self.handles.swap_remove(i);
-            let id = i as u32;
+        while let Some(i) = self.handles.iter().position(Handle::is_unique) {
+            let handle = self.handles.swap_remove(i);
+            let id = handle.id();
             unsafe {
                 self.assets.remove(id);
             }
             self.bitset.remove(id);
-            self.unused_handles.push(old);
+
+            // Can't reuse old handle here, because otherwise weak handles would still be valid.
+            // TODO: maybe just store u32?
```

Comment:
Can reuse again

Reply:
Sorry, I don't understand?

Reply:
You can `Handle`. As it not `WeakHandle` again.

Reply:
I think what you're saying is that because I changed it back to `Handle`, I can reuse the handles again?

I don't think we'd want that, because:

* Handle A gets allocated
* Handle A has a `downgraded` version in a cache
* Handle A gets dropped
* Handle A gets added to the `unused_handles` queue
* Another allocation is required; `unused_handles` gives us the "old" handle A
* The cached value we stored before (using `WeakHandle`) is now valid again

That's why I added that comment and recreated the handle. Did a make a mistake in my reasoning?

Reply:
Makes sense.